src/refract/ExpandVisitor.cc

```diff
@@ -168,35 +162,52 @@ namespace refract
 
         }
 
-#if _MIXIN_EXPANSION_
-        IElement* ExpandReference(const ObjectElement& e, const Registry& registry)
+        template<typename T>
+        T* CreateReference(const IElement& e, const std::string& href, const Registry& registry) {
+
+            typedef T ElementType;
+            ElementType* ref = ExpandMembers(static_cast<const ElementType&>(e), registry, true);
+            ref->element("ref");
+            ref->renderType(e.renderType());
+
+            ElementType* resolved = ExpandNamedType<T>(registry, href, Referenced);
+
+            resolved->renderType(IElement::rFull);
+            ref->attributes["resolved"] = resolved;
+
+            return ref;
+        }
+
+        template <typename T>
+        IElement* ExpandReference(const T& e, const Registry& registry)
         {
-            TypeQueryVisitor tq;
-            StringElement* href = tq.as<StringElement>(FindMemberByKey(e, "href"));
+            StringElement* href = TypeQueryVisitor::as<StringElement>(FindMemberByKey(e, "href"));
 
-            if (href) {
-                IElement* expanded = FindNamedType(registry, href->value);
+            if (!href || href->value.empty()) {
+                return e.clone();
+            }
 
-                if (expanded->empty()) { // if referenced element not found return clone of reference
-                    delete expanded;
-                    expanded = e.clone(); 
-                }
+            IElement* referenced = registry.find(href->value);
+            if (!referenced) {
+                return e.clone();
+            }
+
+            IElement* base = registry.find(e.element());
+            TypeQueryVisitor t;
+            if (base) {
+                base->content(t);
+            }
 
-                return expanded;
+            if (t.get() == TypeQueryVisitor::Array) {
+                return CreateReference<ArrayElement>(e, href->value, registry);
```

Comment:
Oh, you are right. It is not correct. It stays there from original implemententation.
It is working because of same interface ArrayElement and ObjectElement
Will fix